src/bentoml/_internal/runner/runner_handle/remote.py

```diff
@@ -274,6 +274,80 @@ async def async_run_method(
+        *args: P.args,
+        **kwargs: P.kwargs,
+    ) -> t.AsyncGenerator[R, None]:
+        import aiohttp
+
+        from ...runner.container import AutoContainer
+
+        inp_batch_dim = __bentoml_method.config.batch_dim[0]
+
+        headers = {
+            "Bento-Name": component_context.bento_name,
+            "Bento-Version": component_context.bento_version,
+            "Runner-Name": self._runner.name,
+            "Yatai-Bento-Deployment-Name": component_context.yatai_bento_deployment_name,
+            "Yatai-Bento-Deployment-Namespace": component_context.yatai_bento_deployment_namespace,
+        }
+        total_args_num = len(args) + len(kwargs)
+        headers["Args-Number"] = str(total_args_num)
+
+        if total_args_num == 1:
+            # FIXME: also considering kwargs
+            if len(kwargs) == 1:
+                kwarg_name = list(kwargs.keys())[0]
+                headers["Kwarg-Name"] = kwarg_name
+                payload = AutoContainer.to_payload(
+                    kwargs[kwarg_name], batch_dim=inp_batch_dim
+                )
+            else:
+                payload = AutoContainer.to_payload(args[0], batch_dim=inp_batch_dim)
+            data = payload.data
+
+            headers["Payload-Meta"] = json.dumps(payload.meta)
+            headers["Payload-Container"] = payload.container
+            headers["Batch-Size"] = str(payload.batch_size)
+
+        else:
+            payload_params = Params[Payload](*args, **kwargs).map(
+                functools.partial(AutoContainer.to_payload, batch_dim=inp_batch_dim)
+            )
+            data = pickle.dumps(payload_params)  # FIXME: pickle inside pickle
+
+        path = "" if __bentoml_method.name == "__call__" else __bentoml_method.name
+
+        async def stream_response() -> t.AsyncGenerator[R, None]:
+            """
+            Async generator that yields results from the remote runner.
+            """
+            try:
+                async with self._client.post(
+                    f"{self._addr}/{path}",
+                    data=data,
+                    headers=headers,
+                ) as resp:
+                    buffer = bytearray()
+                    async for b, end_of_http_chunk in resp.content.iter_chunks():
+                        buffer.extend(b)
+
+                        # This is to handle large payload that is split into multiple chunks
+                        if end_of_http_chunk and len(buffer) > 0:
+                            # TODO: To remove pickling so that we can stream data as it is
+                            payload = pickle.loads(buffer)
+                            yield AutoContainer.from_payload(payload)
+
+                            # Clearing the buffer for the next data
+                            buffer = bytearray()
+
+            except aiohttp.ClientOSError as e:
+                raise RemoteException("Failed to connect to runner server.") from e
+
+        return stream_response()
```

Comment:
If we just call the function then why do we need to put this into a function in the first place?

Reply:
Because we need to return an async generator, so it's not just executing the function itself.

Reply:
But you call the function immediately right? Wouldn't that be the same as
```python
            try:
                async with self._client.post(
                    f"{self._addr}/{path}",
                    data=data,
                    headers=headers,
jianshen92 marked this conversation as resolved.
                ) as resp:
                    buffer = bytearray()
                    async for b, end_of_http_chunk in resp.content.iter_chunks():
                        buffer.extend(b)

                        # This is to handle large payload that is split into multiple chunks
                        if end_of_http_chunk and len(buffer) > 0:
                            # TODO: To remove pickling so that we can stream data as it is
                            payload = pickle.loads(buffer)
                            yield AutoContainer.from_payload(payload)

                            # Clearing the buffer for the next data
                            buffer = bytearray()

            except aiohttp.ClientOSError as e:
                raise RemoteException("Failed to connect to runner server.") from e
```

Because this itself already returns an async generator?

Reply:
great spot. just updated